### PR TITLE
fix: conversion from hbase to veneer settings

### PR DESCRIPTION
**Pull request by igorbernstein2**
This PR polishes up the conversion from bigtable-hbase `Configuration` settings to bigtable veneer settings.
Major changes:
- timeouts are always used and `google.bigtable.rpc.use.timeouts` is ignored. The operation level timeouts extracted as follows:
  - Scans - in order of preference:`google.bigtable.read.rpc.timeout.ms`, `google.bigtable.long.rpc.timeout.ms`, `google.bigtable.grpc.retry.max.elapsed.backoff.ms`
  - Scans - in order of preference:`google.bigtable.mutate.rpc.timeout.ms`, `google.bigtable.long.rpc.timeout.ms`, `google.bigtable.grpc.retry.max.elapsed.backoff.ms`
  - Everything else - `google.bigtable.rpc.timeout.ms`, `google.bigtable.grpc.retry.max.elapsed.backoff.ms`
- retry and timeout settings no longer apply to admin operations (create table, delete table, etc). The settings will use the defaults from java-bigtable
-

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java [hunk 1]
```diff
@@ -696,17 +650,18 @@ private void configureNonIdempotentCallSettings(UnaryCallSettings.Builder unaryC
...
 
-  private boolean isBatchModeEnabled() {
-    return configuration.getBoolean(BIGTABLE_USE_BATCH, false);
-  }
+    String enableDeadlineRetry = configuration.get(ENABLE_GRPC_RETRY_DEADLINEEXCEEDED_KEY);
+    if (!Strings.isNullOrEmpty(enableDeadlineRetry)) {
+      if (Boolean.parseBoolean(enableDeadlineRetry)) {
+        codes.add(StatusCode.Code.DEADLINE_EXCEEDED);
+      } else {
+        codes.remove(StatusCode.Code.DEADLINE_EXCEEDED);
```
Comment (kolea2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:661]: if DEADLINE_EXCEEDED already there, do we need to add it? (in the if statement)

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestBigtableHBaseVeneerSettings.java [hunk 2]
```diff
@@ -358,17 +334,10 @@ public void testAdminSettingsWithEmulator() throws IOException {
...
   public void testRpcMethodWithoutRetry() throws IOException {
     configuration.set(ENABLE_GRPC_RETRIES_KEY, "false");
-    BigtableHBaseVeneerSettings settingUtils =
-        ((BigtableHBaseVeneerSettings) BigtableHBaseVeneerSettings.create(configuration));
+    BigtableHBaseVeneerSettings settingUtils = BigtableHBaseVeneerSettings.create(configuration);
     BigtableDataSettings settings = settingUtils.getDataSettings();
 
     assertTrue(settings.getStubSettings().readRowsSettings().getRetryableCodes().isEmpty());
-
-    assertRpcWithoutTimeout(settings.getStubSettings().bulkReadRowsSettings());
```
Comment (kolea2) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestBigtableHBaseVeneerSettings.java:367]: can we add new tests in its place?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java [hunk 3]
```diff
@@ -218,50 +215,69 @@ public BigtableTableAdminSettings getTableAdminSettings() {
...
   // ************** Private Helpers **************
   private BigtableDataSettings buildBigtableDataSettings() throws IOException {
-    BigtableDataSettings.Builder dataBuilder =
-        BigtableDataSettings.newBuilder()
-            .setProjectId(getProjectId())
-            .setInstanceId(getInstanceId());
+    BigtableDataSettings.Builder dataBuilder;
+
+    // Configure the Data connection
+    Optional<String> emulatorEndpoint =
```
Comment (kolea2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:223]: please extract this out (looks like it it is repeated later on)

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java [hunk 4]
```diff
@@ -273,99 +289,106 @@ private void configureMetricsBridge(Builder settings) {
...
+    InstantiatingGrpcChannelProvider.Builder channelProvider =
+        ((InstantiatingGrpcChannelProvider) stubSettings.getTransportChannelProvider()).toBuilder();
 
-    InstantiatingGrpcChannelProvider.Builder channelBuilder = defaultGrpcTransportProviderBuilder();
-
-    if (Boolean.parseBoolean(configuration.get(BIGTABLE_USE_PLAINTEXT_NEGOTIATION))) {
-      channelBuilder.setChannelConfigurator(
+    if (configuration.getBoolean(BIGTABLE_USE_PLAINTEXT_NEGOTIATION, false)) {
+      @SuppressWarnings("rawtypes")
+      final ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder> prevConfigurator =
```
Comment (kolea2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:353]: a little confused here - why are we switching from channelBuilder to channelProvider and why the previous configurator?

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java (hunk 3 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:223]: I couldnt figure out a good way to extract this. The newBuilderForEmulator methods are static so there is no good way to make the helper polymorphic for TableAdminSettings and DataSettings.

I can introduce a one off value class to hold the host/port after parsing, but the combination of the extra class and separate function didnt seem worth it to replace 4 lines of splitting on ':'. It would be something like:

```java
@AutoValue
abstract class HostPort {
  abstract String getHost();
  abstract int getPort();
}
Optional<HostPort> getEmulatorEndpint() {
  Optional<String> emulatorEndpoint = Optional.fromNullable(configuration.get(BIGTABLE_EMULATOR_HOST_KEY));
  if (emulatorEndpoint.isPresent()) {
      int split = emulatorEndpoint.get().lastIndexOf(':');
      String host = emulatorEndpoint.get().substring(0, split);
      int port = Integer.parseInt(emulatorEndpoint.get().substring(split + 1));
      return new AutoValue_HostPort(host, port);
    }
return Optional.absent();
}

private BigtableDataSettings buildBigtableDataSettings() {
  Optional<HostPort> emulatorEndpoint = getEmulatorEndpint();
  if (emultorEndpoint.isPresent()) { 
     dataBuilder = BigtableDataSettings.newBuilderForEmulator(emulatorEndpoint.getHost(), emulatorEndpoint.getPort());
  } else {
    dataBuilder = BigtableDataSettings.newBuilder();
      configureConnection(dataBuilder.stubSettings(), BIGTABLE_HOST_KEY);
      configureCredentialProvider(dataBuilder.stubSettings());
 }
 ...
}

```


I guess I can push the configureCredentialProvider & configureHeaderProvider into configureConnection, but I dont think it will help much here. Any suggestions?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java (hunk 1 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:661]: not sure what you mean? Can you rephrase?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestBigtableHBaseVeneerSettings.java (hunk 2 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestBigtableHBaseVeneerSettings.java:367]: Not sure what you mean? This used to test that timeouts were forced to 12 hours. Are you saying that I should add a test that disabling retries doesn't affect timeouts?

Comment (kolea2) [reply bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestBigtableHBaseVeneerSettings.java:367]: yes that would be good

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java (hunk 3 above)
Comment (kolea2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:223]: Maybe a utility class? `buildBigtableDataSettings` might make more sense to leave as is, but emulator config and setup is pretty generic?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java (hunk 1 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:661]: I didnt want to make assumptions about veneer defaults here.

**Review by kolea2 — APPROVED** [review]
